src/storage/lock_manager/lock_waiting_queue.rs

```diff
@@ -0,0 +1,997 @@
+//! [`LockWaitQueues::delayed_notify_all`] will be made. However, since the
+//! [`LockWaitQueues`] do not have its own thread pool, the user may receive a
+//! future after calling some of the functions, and the user will be responsible
+//! for executing the future in a suitable place.
+
+use std::{
+    collections::BinaryHeap,
+    convert::TryFrom,
+    future::Future,
+    pin::Pin,
+    result::Result,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+
+use dashmap;
+use futures_util::compat::Future01CompatExt;
+use kvproto::kvrpcpb;
+use smallvec::SmallVec;
+use sync_wrapper::SyncWrapper;
+use thiserror::Error;
+use tikv_util::{time::InstantExt, timer::GLOBAL_TIMER_HANDLE};
+use txn_types::{Key, TimeStamp};
+
+use crate::storage::{
+    lock_manager::{lock_wait_context::LockWaitContextSharedState, LockManager, LockWaitToken},
+    mvcc::{Error as MvccError, ErrorInner as MvccErrorInner},
+    txn::Error as TxnError,
+    types::{PessimisticLockParameters, PessimisticLockRes},
+    Error as StorageError, ErrorInner as StorageErrorInner,
+};
+
+/// The shared version of [`crate::storage::Error`]. It's necessary to pass a
+/// single error to more than one requests, since the inner error doesn't
+/// support cloning.
+#[derive(Debug, Error)]
+#[error(transparent)]
+pub struct SharedError(Arc<StorageErrorInner>);
+
+impl From<StorageErrorInner> for SharedError {
+    fn from(e: StorageErrorInner) -> Self {
+        Self(Arc::new(e))
+    }
+}
+
+impl From<StorageError> for SharedError {
+    fn from(e: StorageError) -> Self {
+        Self(Arc::from(e.0))
+    }
+}
+
+/// Tries to convert the shared error to owned one. It can success only when
+/// it's the only reference to the error.
+impl TryFrom<SharedError> for StorageError {
+    type Error = ();
+
+    fn try_from(e: SharedError) -> Result<Self, Self::Error> {
+        Arc::try_unwrap(e.0).map(Into::into).map_err(|_| ())
+    }
+}
+
+pub type CallbackWithSharedError<T> = Box<dyn FnOnce(Result<T, SharedError>) + Send + 'static>;
+pub type PessimisticLockKeyCallback = CallbackWithSharedError<PessimisticLockRes>;
+
+/// Represents an `AcquirePessimisticLock` request that's waiting for a lock,
+/// and contains the request's parameters.
+pub struct LockWaitEntry {
+    pub key: Key,
+    pub lock_hash: u64,
+    // TODO: Use term to filter out stale entries in the queue.
+    // pub term: Option<NonZeroU64>,
+    pub parameters: PessimisticLockParameters,
+    pub lock_wait_token: LockWaitToken,
+    pub req_states: Option<Arc<LockWaitContextSharedState>>,
+    pub current_legacy_wake_up_cnt: Option<usize>,
+    pub key_cb: Option<SyncWrapper<PessimisticLockKeyCallback>>,
+}
+
+/// A wrapper to [`LockWaitEntry`] to define its ordering.
+/// A [`LockWaitEntry`] takes precedence over another if it has smaller
+/// `start_ts`.
+#[repr(transparent)]
+struct LockWaitEntryComparableWrapper(pub Box<LockWaitEntry>);
```

Comment:
Yes, if you think it would be better. I was trying to hide the comparison definition, since I think it doesn't make sense to compare them outside the module.

Reply:
Ah, I see. You want it only comparable in this module. I think it makes sense although the name looks a bit too long. `ComparableLockWaitEntry` may be a bit shorter. And `into_inner` is the more common name for than `unwrap` for extracting the inner object.